<a href="https://colab.research.google.com/github/Jihyesh/codestates_projects_section1/blob/main/AI_15_%ED%99%8D%EC%A7%80%ED%98%9C_Section1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


In [ ]:
from google.colab import files
files.upload()

In [3]:
df = pd.read_csv('vgames2.csv')
df.head()

Unnamed: 0                                            Name Platform  \
0           1                    Candace Kane's Candy Factory       DS   
1           2                                  The Munchables      Wii   
2           3  Otome wa Oanesama Boku ni Koi Shiteru Portable      PSP   
3           4                Deal or No Deal: Special Edition       DS   
4           5       Ben 10 Ultimate Alien: Cosmic Destruction      PS3   

     Year      Genre           Publisher NA_Sales EU_Sales JP_Sales  \
0  2008.0     Action           Destineer     0.04        0        0   
1  2009.0     Action  Namco Bandai Games     0.17        0        0   
2  2010.0  Adventure           Alchemist        0        0     0.02   
3  2010.0       Misc           Zoo Games     0.04        0        0   
4  2010.0   Platform         D3Publisher     0.12     0.09        0   

  Other_Sales  
0           0  
1        0.01  
2           0  
3           0  
4        0.04

###Mission
<font color='blue'>비데이터 직군</font>의 청자에게 데이터를 기반으로 위의 주제에 대해 설명하는 영상 제출
- **지역에 따라서 선호하는 게임 장르가 다를까** 라는 질문에 대답을 하셔야 합니다
- **연도별 게임의 트렌드가 있을까** 라는 질문에 대답을 하셔야 합니다
- **출고량의 높은 게임에 대한 분석 및 시각화 프로세스**가 포함되어야 합니다
- **결론적으로 다음 분기에 어떤 게임을 설계해야 하는지를 제시**하셔야 합니다
- 추가 인사이트


###EDA

In [18]:
df.shape

(16598, 10)

In [42]:
df.dtypes

Unnamed: 0       int64
Name            object
Platform        object
Year           float64
Genre           object
Publisher       object
NA_Sales        object
EU_Sales        object
JP_Sales        object
Other_Sales     object
dtype: object

In [27]:
df.sample(10)

Unnamed: 0                                               Name Platform  \
14912       14913                                  The Wonderful 101     WiiU   
6944         6945            Professor Layton and the Unwound Future       DS   
11298       11299                       Harukanaru Toki no Naka de 3      PS2   
4015         4016      Samantha Swift and the Hidden Roses of Athena       DS   
3209         3210                        Hyperdimension Neptunia mk2      PSV   
1684         1685  SpongeBob SquarePants featuring Nicktoons: Glo...      PS2   
6927         6928                 SBK09 Superbike World Championship      PS3   
4632         4633            Naruto: Uzumaki Chronicles 2 (JP sales)      PS2   
13242       13243                   Aidyn Chronicles: The First Mage      N64   
13883       13884                                    Haunting Ground      PS2   

         Year         Genre           Publisher NA_Sales EU_Sales JP_Sales  \
14912  2013.0        Action            Nintendo      0.2      0.1     0.08   
6944   2008.0        Puzzle            Nintendo     0.65     1.61     0.82   
11298  2004.0     Adventure          Tecmo Koei        0        0     0.05   
4015   2010.0     Adventure             Unknown     0.04     0.01        0   
3209   2014.0        Action       Compile Heart        0        0     0.04   
1684      8.0        Action                 THQ     0.29     0.22        0   
6927   2009.0        Racing    Black Bean Games        0     0.01        0   
4632   2006.0        Action  Namco Bandai Games        0        0     0.05   
13242  2001.0  Role-Playing                 THQ     0.08     0.02        0   
13883  2005.0        Action              Capcom    0.03M     0.03        0   

      Other_Sales  
14912        0.03  
6944         0.28  
11298           0  
4015            0  
3209            0  
1684         0.07  
6927            0  
4632            0  
13242           0  
13883        0.01

In [12]:
df.isna().sum()

Unnamed: 0       0
Name             0
Platform         0
Year           271
Genre           50
Publisher       58
NA_Sales         0
EU_Sales         0
JP_Sales         0
Other_Sales      0
dtype: int64

####EDA-Year

In [13]:
df.Year.dtype

dtype('float64')

In [17]:
# Year missing value
df[df.Year.isnull()]

Unnamed: 0                                         Name Platform  Year  \
31             32                                    Rock Band     X360   NaN   
109           110   Yu-Gi-Oh! 5D's Wheelie Breakers (JP sales)      Wii   NaN   
273           274  Jewel Link Chronicles: Mountains of Madness       DS   NaN   
358           359           Majesty 2: The Fantasy Kingdom Sim     X360   NaN   
430           431                                   Famista 64      N64   NaN   
...           ...                                          ...      ...   ...   
16240       16241                                Freaky Flyers       GC   NaN   
16275       16276                               Space Invaders     2600   NaN   
16278       16279                                Madden NFL 11      Wii   NaN   
16425       16426                               Yoostar on MTV     X360   NaN   
16432       16433                                  Bejeweled 3     X360   NaN   

            Genre           Publisher NA_Sales EU_Sales JP_Sales Other_Sales  
31           Misc     Electronic Arts     1.93     0.34        0        0.21  
109        Racing             Unknown        0        0     0.02           0  
273        Puzzle           Avanquest        0     0.06        0        0.01  
358    Simulation             Unknown     0.03        0        0           0  
430        Sports  Namco Bandai Games        0        0     0.17        0.03  
...           ...                 ...      ...      ...      ...         ...  
16240      Racing             Unknown     0.01        0        0           0  
16275     Shooter               Atari     2.36     0.14       0M        0.03  
16278      Sports             Unknown      0.7        0        0         50K  
16425        Misc             Unknown     0.06        0        0           0  
16432      Puzzle             Unknown     0.13        0        0        0.01  

[271 rows x 10 columns]

In [19]:
df[df.Year.isnull()].shape[0]/df.shape[0]

0.016327268345583804

####EDA-Platform

In [29]:
df[df.Platform.isnull()]

Empty DataFrame
Columns: [Unnamed: 0, Name, Platform, Year, Genre, Publisher, NA_Sales, EU_Sales, JP_Sales, Other_Sales]
Index: []

In [5]:
df.Platform.unique()

array(['DS', 'Wii', 'PSP', 'PS3', 'PC', 'PS', 'GBA', 'PS4', 'PS2', 'XB',
       'X360', 'GC', '3DS', '2600', 'SAT', 'GB', 'NES', 'DC', 'N64',
       'XOne', 'SNES', 'WiiU', 'PSV', 'GEN', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

In [6]:
df.Platform.nunique()

31

####EDA-Genre

In [31]:
df[df.Genre.isnull()].shape[0]

50

In [32]:
df[df.Genre.isnull()]

Unnamed: 0                                         Name Platform  \
50             51                     Ape Escape: On the Loose      PSP   
408           409                                  NBA Ballers       XB   
1328         1329                                   NFL Street       XB   
1582         1583               SingStar Singalong With Disney      PS2   
1971         1972                          Pokemon X/Pokemon Y      3DS   
2043         2044                                NASCAR Racing       PS   
2497         2498             Jurassic Park III: Island Attack      GBA   
3353         3354                  The Godfather: Dons Edition      PS3   
3363         3364                 Ski-Doo Snowmobile Challenge     X360   
3470         3471           Batman Beyond: Return of the Joker      N64   
3515         3516     Princess Saver! Eternal Love for My Lady      PS2   
3932         3933                            MVP Baseball 2004      PS2   
4148         4149                          Grand Theft Auto IV       PC   
4349         4350                    Brain Quest: Grades 5 & 6       DS   
4545         4546            Dragon Ball Z: Ultimate Tenkaichi     X360   
4598         4599                                    Wii Party      Wii   
4935         4936                             Astonishia Story      PSP   
5668         5669                                 Chrono Cross       PS   
5713         5714                 Back to the Future: The Game     XOne   
5945         5946                            Medabots: Metabee      GBA   
6798         6799                                 UFO: Trilogy       PC   
7143         7144                 Largo Winch .// Commando Sar       PS   
7336         7337                                Madden NFL 11      PS2   
7526         7527                              Ukiyo no Shishi      PS3   
7797         7798                    Myth Makers: Orbs of Doom      Wii   
9488         9489                            Final Fantasy XII      PS2   
10432       10433        The Mummy: Tomb of the Dragon Emperor      Wii   
10501       10502                      Bakemonogatari Portable      PSP   
10874       10875            Karaoke Revolution Glee: Volume 3      Wii   
11050       11051                                 Puppy Palace       DS   
11063       11064                                     Yakuza 3      PS3   
11268       11269            Cloudy With a Chance of Meatballs      PS3   
11496       11497                                     F-Zero X      N64   
11668       11669  The History Channel: Battle for the Pacific      Wii   
11699       11700                               Time Travelers      PSV   
12314       12315                                   Ice Hockey     2600   
12331       12332                           Onimusha: Warlords      PS2   
12340       12341                             Virtual Chess 64      N64   
12870       12871         PQ2: Practical Intelligence Quotient      PSP   
12982       12983                                  Urban Chaos       PS   
13348       13349        Treasure Report: Kikai Jikake no Isan       DS   
14930       14931                            Skylanders Giants      PS3   
15077       15078                   Myth Makers: Super Kart GP      Wii   
15211       15212                    James Bond 007: Nightfire       XB   
15426       15427            WarioWare, Inc.: Mega Party Game$       GC   
15526       15527                                   Bloodborne      PS4   
16091       16092                       Brain Boost: Beta Wave       DS   
16431       16432            Cars: Mater-National Championship     X360   
16555       16556                          Apache: Air Assault       PC   
16587       16588                 Need for Speed Underground 2       XB   

         Year Genre                     Publisher NA_Sales EU_Sales JP_Sales  \
50     2005.0   NaN   Sony Computer Entertainment     0.52     0.01     0.12   
408    2004.0   NaN  

In [33]:
df.Genre.unique()

array(['Action', 'Adventure', 'Misc', 'Platform', 'Sports', 'Simulation',
       'Racing', 'Role-Playing', 'Puzzle', 'Strategy', 'Fighting',
       'Shooter', nan], dtype=object)

In [34]:
df[df.Genre == 'Platform']

Unnamed: 0                                       Name Platform    Year  \
4               5  Ben 10 Ultimate Alien: Cosmic Destruction      PS3  2010.0   
20             21                     Ty the Tasmanian Tiger      PS2  2002.0   
25             26                                   Dr. Muto       GC  2002.0   
28             29                    Famicom Mini: Makaimura      GBA  2004.0   
34             35            Mega Man Anniversary Collection      PS2  2004.0   
...           ...                                        ...      ...     ...   
16531       16532                  New Super Mario Bros. Wii      Wii  2009.0   
16563       16564                                     Wall-E      Wii  2008.0   
16578       16579                       Gex: Enter the Gecko       PS  1998.0   
16579       16580                     Super Ghouls 'n Ghosts      GBA  2002.0   
16593       16594                    Ice Age 2: The Meltdown       GC  2006.0   

          Genre                      Publisher NA_Sales EU_Sales JP_Sales  \
4      Platform                    D3Publisher     0.12     0.09        0   
20     Platform                Electronic Arts     0.57     0.45        0   
25     Platform                   Midway Games     0.02     0.01        0   
28     Platform                         Capcom        0        0     0.08   
34     Platform                         Capcom     0.29     0.22        0   
...         ...                            ...      ...      ...      ...   
16531  Platform                       Nintendo    14.59     7.06      4.7   
16563  Platform                            THQ     0.51      0.4        0   
16578  Platform  BMG Interactive Entertainment     320K     0.22        0   
16579  Platform                         Capcom     0.05     0.02        0   
16593  Platform                  Vivendi Games     0.15     0.04        0   

      Other_Sales  
4            0.04  
20           0.15  
25              0  
28              0  
34           0.08  
...           ...  
16531        2.26  
16563        0.11  
16578        0.04  
16579           0  
16593        0.01  

[884 rows x 10 columns]

####EDA-PUblisher

In [36]:
df[df.Publisher.isnull()]

Unnamed: 0                                               Name Platform  \
272           273                                 UK Truck Simulator       PC   
1313         1314     All Grown Up!: Game Boy Advance Video Volume 1      GBA   
1560         1561  Teenage Mutant Ninja Turtles: Game Boy Advance...      GBA   
1789         1790  Nicktoons Collection: Game Boy Advance Video V...      GBA   
1879         1880                     Monster Hunter Frontier Online      PS3   
1961         1962  Nicktoons Collection: Game Boy Advance Video V...      GBA   
2070         2071                                     World of Tanks     X360   
2970         2971                               Chou Soujuu Mecha MG       DS   
3184         3185                       Action Man-Operation Extreme       PS   
3412         3413                                  Dance with Devils      PSV   
3533         3534                          B.L.U.E.: Legend of Water       PS   
3562         3563                         wwe Smackdown vs. Raw 2006      PS2   
3639         3640                    Homeworld Remastered Collection       PC   
3684         3685    Dragon Ball GT: Game Boy Advance Video Volume 1      GBA   
4308         4309  Umineko no Naku Koro ni San: Shinjitsu to Gens...      PS3   
4367         4368  Strawberry Shortcake: Game Boy Advance Video V...      GBA   
5207         5208  Cartoon Network Collection: Game Boy Advance V...      GBA   
5245         5246  Cartoon Network Collection: Game Boy Advance V...      GBA   
5387         5388  Moshi, Kono Sekai ni Kami-sama ga Iru to suru ...      PSV   
6209         6210             Atsumare! Power Pro Kun no DS Koushien       DS   
6392         6393                        Brothers in Arms: Furious 4     X360   
7370         7371                                       Stronghold 3       PC   
7984         7985                   Disney Channel Collection Vol. 1      GBA   
8038         8039  Super Robot Monkey Team: Game Boy Advance Vide...      GBA   
8236         8237                                 Bentley's Hackpack      GBA   
8355         8356  Cartoon Network Collection: Game Boy Advance V...      GBA   
8651         8652                                 Sonic the Hedgehog      PS3   
8873         8874                    Legends of Oz: Dorothy's Return      3DS   
9128         9129  The Fairly Odd Parents: Game Boy Advance Video...      GBA   
9684         9685      Dragon Ball Z: Budokai Tenkaichi 2 (JP sales)      Wii   
9877         9878                                       Mario Tennis      3DS   
10156       10157  The Legend of Zelda: The Minish Cap(weekly JP ...      GBA   
10223       10224                   Demolition Company: Gold Edition       PC   
11309       11310                                     Triple Play 99       PS   
11498       11499  Nicktoons Collection: Game Boy Advance Video V...      GBA   
11641       11642             AKB1/48: Idol to Guam de Koishitara...     X360   
11676       11677                             Farming Simulator 2011       PC   
12126       12127                                      Silverlicious       DS   
12176       12177  Dora the Explorer: Game Boy Advance Video Volu...      GBA   
12468       12469  The Fairly Odd Parents: Game Boy Advance Video...      GBA   
12560       12561  SpongeBob SquarePants: Game Boy Advance Video ...      GBA   
12900       12901  SpongeBob SquarePants: Game Boy Advance Video ...      GBA   
13182       13183           Sonic X: Game Boy Advance Video Volume 1      GBA   
13204       13205                                            Xia-Xia       DS   
13450       13451                                Bikkuriman Daijiten       DS   
13766       13767                                       Housekeeping       DS   
13810       13811  Pokémon: Johto Photo Finish: Game Boy Advance ...      GBA   
13837       13838  Cartoon Network Collection: Game Boy Advance V...      GBA   
13923       13924       Prinny: Can I 

In [37]:
df.Publisher.unique()

array(['Destineer', 'Namco Bandai Games', 'Alchemist', 'Zoo Games',
       'D3Publisher', 'Konami Digital Entertainment', 'Deep Silver',
       'Tecmo Koei', 'Sega', 'THQ', 'Empire Interactive', 'Ubisoft',
       'Paradox Interactive', '505 Games', 'Tomy Corporation',
       'Eidos Interactive', 'Microsoft Game Studios',
       'Nippon Ichi Software', 'Kalypso Media', 'Electronic Arts',
       'Sony Computer Entertainment', 'Activision',
       'Tru Blu Entertainment', 'Warner Bros. Interactive Entertainment',
       'Midway Games', 'Mindscape', 'Wargaming.net', 'Capcom',
       'Acclaim Entertainment', 'Unknown', 'Nintendo',
       'Disney Interactive Studios', 'Crave Entertainment', 'Sunflowers',
       'Arc System Works', 'Milestone S.r.l', 'Kamui', 'Hudson Soft',
       'Square Enix', 'Focus Home Interactive', 'Take-Two Interactive',
       'Game Factory', 'Takara Tomy', 'SquareSoft', 'Global Star',
       'Infogrames', 'GT Interactive', 'Storm City Games',
       'Gathering of Dev

In [40]:
df.Publisher.nunique()

578

####EDA-Sales

In [79]:
df.sample(15)

Unnamed: 0                                               Name Platform  \
6193         6194  Zaidan Houjin Nippon Kanji Nouryoku Kentei Kyo...       DS   
16289       16290                                          Fallout 4     XOne   
15985       15986                                     Sugoi Hebereke     SNES   
14829       14830                                Knockout Kings 2002       XB   
6149         6150                                      Farm Frenzy 3       PC   
15307       15308                                  Fire Emblem Fates      3DS   
3230         3231                                   ATV: Quad Frenzy       DS   
2868         2869                                 Fallout: New Vegas     X360   
630           631          Triple Pack: Xbox Live Arcade Compilation     X360   
8366         8367                                        NBA Live 15      PS4   
4625         4626                      TimeSplitters: Future Perfect      PS2   
8969         8970                         Dave Mirra Freestyle BMX 2      PS2   
5533         5534                                Tigger's Honey Hunt       PS   
3815         3816                      Asterix & Obelix: Kick Buttix      PS2   
1060         1061                                          King Kong     2600   

         Year         Genre               Publisher NA_Sales EU_Sales  \
6193   2007.0          Misc          Rocket Company        0        0   
16289  2015.0  Role-Playing      Bethesda Softworks     2.45     1.26   
15985    94.0      Fighting                 Sunsoft        0        0   
14829  2002.0        Sports         Electronic Arts      0.2     0.06   
6149   2009.0      Strategy    Alawar Entertainment        0     0.03   
15307  2015.0  Role-Playing                Nintendo     0.71     0.22   
3230   2005.0        Racing   Majesco Entertainment      0.1        0   
2868   2010.0  Role-Playing      Bethesda Softworks     2.65     1.06   
630    2011.0          Misc  Microsoft Game Studios     0.03        0   
8366   2014.0        Sports         Electronic Arts     0.15     0.04   
4625   2005.0       Shooter         Electronic Arts     0.12     0.09   
8969   2001.0        Sports   Acclaim Entertainment     0.67     0.53   
5533   2000.0      Platform                 Ubisoft     0.16     0.11   
3815   2004.0        Action                   Atari     0.02     0.01   
1060   1981.0        Action             Tigervision     0.65     0.04   

      JP_Sales Other_Sales  
6193      0.33           0  
16289     0.01        0.37  
15985     0.02           0  
14829        0        0.01  
6149         0        0.01  
15307     0.52         0.1  
3230         0        0.01  
2868      0.04        0.33  
630          0           0  
8366         0        0.04  
4625         0        0.03  
8969         0        0.18  
5533         0        0.02  
3815         0           0  
1060         0        0.01

In [83]:
# K, M 있는 것만 꺼내보자
# foo[foo.b.str.contains('oo', regex= True, na=False)]

# df.columns
# 'NA_Sales', 'EU_Sales', 'JP_Sales', 'Other_Sales'

df[df.NA_Sales.str.contains('M',regex=True, na=False)]

Unnamed: 0                                        Name Platform  \
447           448                  Medal of Honor: Warfighter     X360   
458           459                                     F1 2010      PS3   
577           578                               Mario Kart 64      N64   
625           626                                  The Hobbit      GBA   
672           673  Star Wars The Clone Wars: Lightsaber Duels      Wii   
...           ...                                         ...      ...   
15986       15987                 Juiced 2: Hot Import Nights       PC   
16152       16153                       Taiko no Tatsujin Wii      Wii   
16169       16170                    Medal of Honor: Airborne     X360   
16408       16409               NASCAR 06: Total Team Control      PS2   
16482       16483            Super Robot Taisen F Kanketsuhen      SAT   

         Year     Genre           Publisher NA_Sales EU_Sales JP_Sales  \
447    2012.0    Action     Electronic Arts    0.58M     0.47     0.01   
458    2010.0    Racing         Codemasters    0.25M     0.76     0.08   
577    1996.0    Racing            Nintendo    5.55M     1.94     2.23   
625    2003.0  Platform       Vivendi Games    0.14M     0.05        0   
672    2008.0    Action           LucasArts    1.23M      0.3        0   
...       ...       ...                 ...      ...      ...      ...   
15986  2007.0    Racing                 THQ       0M     0.01        0   
16152  2008.0      Misc  Namco Bandai Games       0M        0     0.67   
16169  2007.0   Shooter     Electronic Arts    0.45M     0.53     0.01   
16408  2005.0    Racing     Electronic Arts    0.53M     0.02        0   
16482    98.0  Strategy           Banpresto       0M        0     0.55   

      Other_Sales  
447           0.1  
458          0.27  
577          0.15  
625             0  
672          0.14  
...           ...  
15986           0  
16152           0  
16169        0.12  
16408        0.09  
16482           0  

[100 rows x 10 columns]

In [84]:
df[df.NA_Sales.str.contains('K',regex=True, na=False)]

Unnamed: 0                                            Name Platform  \
10             11                                     Ford Racing       PS   
45             46                    Resident Evil: Revelations 2      PS3   
145           146                              TOCA Race Driver 3       PC   
878           879  Tears to Tiara Anecdotes: The Secret of Avalon      PS3   
1086         1087                       Pro Evolution Soccer 2014      3DS   
...           ...                                             ...      ...   
15647       15648                       Championship Manager 2010       PC   
16194       16195                            Vin Diesel: Wheelman      PS3   
16330       16331                         Midway Arcade Treasures      PS2   
16492       16493                                   Mortal Kombat      PSV   
16578       16579                            Gex: Enter the Gecko       PS   

         Year     Genre                               Publisher NA_Sales  \
10     2001.0    Racing                      Empire Interactive     480K   
45     2015.0    Action                                  Capcom      60K   
145    2006.0    Racing                             Codemasters       0K   
878    2009.0  Strategy                               Aqua Plus       0K   
1086   2013.0    Action            Konami Digital Entertainment       0K   
...       ...       ...                                     ...      ...   
15647  2009.0    Sports                       Eidos Interactive       0K   
16194  2009.0    Racing                                 Ubisoft     120K   
16330  2003.0      Misc                            Midway Games     720K   
16492  2012.0  Fighting  Warner Bros. Interactive Entertainment     470K   
16578  1998.0  Platform           BMG Interactive Entertainment     320K   

      EU_Sales JP_Sales Other_Sales  
10       0.33M       0K        0.06  
45        0.08     0.16        0.03  
145       0.02        0           0  
878          0     0.04           0  
1086         0     0.19           0  
...        ...      ...         ...  
15647     0.01        0           0  
16194     0.15        0        0.06  
16330     0.56        0        0.19  
16492     0.18        0        0.12  
16578     0.22        0        0.04  

[100 rows x 10 columns]

In [87]:
df[df.iloc[:,-1].str.contains('M',regex=True, na=False)]

Unnamed: 0                                               Name Platform  \
64             65                                    SD Gundam Force      GBA   
73             74                       Teenage Mutant Ninja Turtles       XB   
209           210              Harry Potter and the Sorcerer's Stone       XB   
220           221          Dragon Quest IV: Michibikareshi Monotachi       PS   
243           244                 Seaman 2: Peking Genjin Ikusei Kit      PS2   
...           ...                                                ...      ...   
15998       15999                          Pro Yakyuu Family Stadium      NES   
16036       16037                                      Point Blank 2       PS   
16074       16075                     The Black Eyed Peas Experience      Wii   
16141       16142  World Championship Poker: Featuring Howard Led...      PS2   
16412       16413  Yu-Gi-Oh! Ultimate Masters: World Championship...      GBA   

         Year         Genre                     Publisher NA_Sales EU_Sales  \
64     2004.0       Shooter            Namco Bandai Games     0.01        0   
73     2003.0        Action  Konami Digital Entertainment     0.22     0.06   
209    2003.0        Action               Electronic Arts     0.05     0.02   
220    2001.0  Role-Playing              Enix Corporation        0        0   
243    2007.0    Simulation                          Sega        0        0   
...       ...           ...                           ...      ...      ...   
15998  1986.0        Sports            Namco Bandai Games     0.15        0   
16036  1998.0       Shooter            Namco Bandai Games     0.02     0.02   
16074  2011.0          Misc                       Ubisoft     0.39      0.2   
16141  2006.0          Misc            Oxygen Interactive     0.82     0.64   
16412  2006.0          Misc  Konami Digital Entertainment     0.14     0.05   

      JP_Sales Other_Sales  
64           0          0M  
73           0       0.01M  
209          0          0M  
220        1.2       0.01M  
243       0.06          0M  
...        ...         ...  
15998     2.05          0M  
16036        0          0M  
16074        0       0.06M  
16141        0       0.22M  
16412     0.07       0.01M  

[100 rows x 10 columns]

In [113]:
df_test = df.copy()

#####not working

Delete non-number in characters using regular expression

`re.sub(pattern, new, string)`<br>
string에서 pattern에 해당하는 부분을 new로 대체하라는 의미 <br>
(replace 'pattern' with 'new' in 'string')

```python
# 함수를 만들어 구현
import re

def convert(phone_number):
  phone_number = re.sub(r"/D","", phone_number)[-8:]   #\D = [^0-9] \d = [0-9]
  final_number = '010' + '-' + phone_number[:4] + '-' + phone_number[-4:]
  return final_number
  
# 만든 함수를 .apply() 함수를 이용해 적용
patients_clean.전화번호 = patients_clean.전화번호.apply(convert)
```

[pandas.DataFrame.astype](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.astype.html)
- DataFrame.astype(dtype, copy=True, errors='raise')
- Cast a pandas object to a specified dtype dtype. <pr>

[Column에서 astype을 쓰는 방법](https://cosmosproject.tistory.com/123)
```python
df_test.loc[:, ['col1']] = df_test.loc[:, ['col1']].astype('str')
```

In [96]:
import re

def convert(sales_figure):
  sales_figure = re.sub(r"\D","",sales_figure)
  # sales_figure = sales_figure.astype(float)
  return sales_figure

[range(start, stop, step)](https://www.geeksforgeeks.org/python-range-function/)

- start: integer starting from which the sequence of integers is to be returned
- stop: integer before which the sequence of integers is to be returned. The range of integers ends at stop – 1.
- step: integer value which determines the increment between each integer in the sequence

In [94]:
for i in range(1,5,1):
  df_test.iloc[:,-i] = df_test.iloc[:,-i].apply(convert)

AttributeError: ignored

In [95]:
for i in range(1,5,1):
  df_test.iloc[:,-i] = df_test.iloc[:,-i].astype(float)

ValueError: ignored

#####working

In [115]:
import re

def convert(sales_figure):
  sales_figure = re.sub(r"\D","",sales_figure)

  return sales_figure

In [97]:
df_test.iloc[:,-1]=df_test.iloc[:,-1].apply(convert)

In [114]:
df_test.iloc[[64]]

Unnamed: 0             Name Platform    Year    Genre           Publisher  \
64          65  SD Gundam Force      GBA  2004.0  Shooter  Namco Bandai Games   

   NA_Sales EU_Sales JP_Sales Other_Sales  
64     0.01        0        0          0M

In [104]:
for i in range(1,5,1):
  df_test.iloc[:,-i] = df_test.iloc[:,-i].apply(convert)

#####not working

In [ ]:
index_list = df[df.NA_Sales.str.contains('M',regex=True, na=False) | df.EU_Sales.str.contains('M',regex=True, na=False) | \
                df.JP_Sales.str.contains('M',regex=True, na=False) | df.Other_Sales.str.contains('M',regex=True, na=False) ].index.tolist()
index_list

In [112]:
df_test[[index_list,:]]

SyntaxError: ignored

#####working

**First of all, `search` for the rows to be fixed. This process results in a list of True/False**

```python
import re

def has_errors(inputString):
  return bool(re.search('[_,!,.,\d+]', inputString))
  
  # (설명)re.search --- inputString arguments에 _ ! . \d+ 숫자(한 개 이상 +) 찾으면 
  # (설명) bool --- True/False로 나온다
```

```python
# 이름에 섞여 있는 오류 데이터를 확인
error_condition = patients_clean.이름.apply(has_errors)
error_names = patients_clean[error_condition]
error_names

 # (설명) error_condition에 어떤 조건을 넣고, 그 조건이 맞는 데이터프레임을 error_name으로 불러옴
 ```

 
**Second, `replace` characters in values of the rows that has been searched in the first step**
```python
# 코드가 정확하게 오류를 제거하는지 확인
error_names.이름 = error_names.이름.str.replace('[_,!,.,\d+]','')
error_names

# (설명) str.replace()는 시리즈에서 _,!,.,\d+를 없애는 것으로 대체함
```
**Finally, apply the same `str.replace` method to the target DataFrame**
```python
# 제거된 것을 확인하였으면, 환자데이터 이름 컬럼에 적용
patients_clean.이름 = patients_clean.이름.str.replace('[_,!,.,\d+]','')
```

In [124]:
import re

def has_errors(inputString):
  return bool(re.search('[M,K]',inputString))

In [129]:
# call list of boolean that indicates number(False) or non-number(True)
not_num_list_NA = df_test.NA_Sales.apply(has_errors)
not_num_list_EU = df_test.EU_Sales.apply(has_errors)
not_num_list_JP = df_test.JP_Sales.apply(has_errors)
not_num_list_Other = df_test.Other_Sales.apply(has_errors)

In [131]:
df_test_figure_err = df_test[not_num_list_NA | not_num_list_EU | not_num_list_JP | not_num_list_Other]
df_test_figure_err

Unnamed: 0                                   Name Platform    Year  \
10             11                            Ford Racing       PS  2001.0   
36             37              Advance Wars: Dual Strike       DS  2005.0   
45             46           Resident Evil: Revelations 2      PS3  2015.0   
64             65                        SD Gundam Force      GBA  2004.0   
73             74           Teenage Mutant Ninja Turtles       XB  2003.0   
...           ...                                    ...      ...     ...   
16525       16526                        Planet Monsters      GBA  2001.0   
16571       16572                           Koihime Enbu      PS4  2016.0   
16578       16579                   Gex: Enter the Gecko       PS  1998.0   
16580       16581  Captain Morgane and the Golden Turtle       DS  2012.0   
16591       16592                        Madden NFL 2005      PS2  2004.0   

           Genre                      Publisher NA_Sales EU_Sales JP_Sales  \
10        Racing             Empire Interactive     480K    0.33M       0K   
36      Strategy                       Nintendo      0.3    0.03M     0.04   
45        Action                         Capcom      60K     0.08     0.16   
64       Shooter             Namco Bandai Games     0.01        0        0   
73        Action   Konami Digital Entertainment     0.22     0.06        0   
...          ...                            ...      ...      ...      ...   
16525     Action                          Titus     0.01       0K        0   
16571   Fighting                           Yeti        0       0M     0.02   
16578   Platform  BMG Interactive Entertainment     320K     0.22        0   
16580  Adventure             Reef Entertainment        0     0.03       0M   
16591     Sports                Electronic Arts     4.18     0.26     0.01   

      Other_Sales  
10           0.06  
36           0.03  
45           0.03  
64             0M  
73          0.01M  
...           ...  
16525           0  
16571           0  
16578        0.04  
16580           0  
16591         80K  

[785 rows x 10 columns]

In [134]:
# df_test_figure_err.iloc[:,-4:-1] = df_test_figure_err.iloc[:,-4:-1].str.replace("[M,K]","")
# DataFrame에선 str 안 돌아가니까, column단위에서 보기 위해서 for문 돌림

for i in range(1,5,1):
  df_test_figure_err.iloc[:,-i] = df_test_figure_err.iloc[:,-i].str.replace("[M,K]","")

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: FutureWarning: The default value of regex will change from True to False in a future version.
  after removing the cwd from sys.path.
/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:1773: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


In [156]:
df_test_figure_err.iloc[:,0:10]
df_test_figure_err.iloc[:,6:10]
df_test_figure_err.iloc[:,-4:-1] #? 왜 마지막이 안 나오지..

NA_Sales EU_Sales JP_Sales
10         480     0.33        0
36         0.3     0.03     0.04
45          60     0.08     0.16
64        0.01        0        0
73        0.22     0.06        0
...        ...      ...      ...
16525     0.01        0        0
16571        0        0     0.02
16578      320     0.22        0
16580        0     0.03        0
16591     4.18     0.26     0.01

[785 rows x 3 columns]

####EDA-Duplicated

In [90]:
df[df.duplicated()]

Empty DataFrame
Columns: [Unnamed: 0, Name, Platform, Year, Genre, Publisher, NA_Sales, EU_Sales, JP_Sales, Other_Sales]
Index: []

####EDA result
- Null: year 271, Genre 50, Publisher 58
- Year data type : float to integer, drop 271 null rows (1.6% of raw data)
- Sales : unit alignment ex. 0M, 50K
- Genre : drop 50 null rows
- Publisher : drop 58 null rows
- Sales : K,M